src/main/java/net/minestom/server/entity/type/projectile/EntityAbstractArrow.java

```diff
@@ -0,0 +1,161 @@
+import net.minestom.server.utils.Vector;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+import java.util.Collection;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+public class EntityAbstractArrow extends ObjectEntity implements Projectile {
+
+    private final static byte CRITICAL_BIT = 0x01;
+    private final static byte NO_CLIP_BIT  = 0x02;
+
+    private final Entity shooter;
+
+    EntityAbstractArrow(@Nullable Entity shooter, @NotNull EntityType entityType, @NotNull Position spawnPosition) {
+        super(entityType, spawnPosition);
+        this.shooter = shooter;
+        super.hasPhysics = false;
+
+        setBoundingBox(.5F, .5F, .5F);
+    }
+
+    @Override
+    public void tick(long time) {
+        Position posBefore = getPosition().clone();
+        super.tick(time);
+        Position posNow = getPosition().clone();
+        if (isStuck(posBefore, posNow)) {
+            if (super.onGround) {
+                return;
+            }
+            super.onGround = true;
+            getVelocity().zero();
+            sendPacketToViewersAndSelf(getVelocityPacket());
+            setNoGravity(true);
+        } else {
+            if (!super.onGround) {
+                return;
+            }
+            super.onGround = false;
+            setNoGravity(false);
+        }
+    }
+
+    @SuppressWarnings("ConstantConditions")
+    private boolean isStuck(Position pos, Position posNow) {
+        if (pos.isSimilar(posNow)) {
+            return true;
+        }
+
+        Instance           instance = getInstance();
+        Chunk              chunk    = null;
+        Collection<Entity> entities = null;
+
+        double   part      = .25D; // half of the bounding box
+        Vector   dir       = posNow.toVector().subtract(pos.toVector());
+        int      parts     = (int) Math.ceil(dir.length() / part);
+        Position direction = dir.normalize().multiply(part).toPosition();
+        for (int i = 0; i < parts; ++i) {
+            if (i == parts - 1) {
+                pos.setX(posNow.getX());
+                pos.setY(posNow.getY());
+                pos.setZ(posNow.getZ());
+            } else {
+                pos.add(direction);
+            }
+            BlockPosition bpos  = pos.toBlockPosition();
+            Block         block = getInstance().getBlock(bpos.getX(), bpos.getY() - 1, bpos.getZ());
+            if (!block.isAir() && !block.isLiquid()) {
+                teleport(pos);
+                return true;
+            }
+
+            Chunk currentChunk = instance.getChunkAt(pos);
+            if (currentChunk != chunk) {
+                chunk = currentChunk;
+                entities = instance.getChunkEntities(chunk)
+                        .stream()
+                        .filter(entity -> entity instanceof LivingEntity)
+                        .collect(Collectors.toSet());
+            }
+            if (getAliveTicks() < 3) {
```

Comment:
Could you add a comment as to why it is necessary? (maybe also in other part of the code, as it is pretty hard to get at first glance)

Reply:
Because arrows spawns in the shooter and without this check it will instantly damage him. And we also can't exclude shooter from the targets list, because you shooter should be able to launch an arrow vertically up and then take damage from it.